internal/engine/crd_test.go

```diff
@@ -0,0 +1,264 @@
+			g.Expect(cclist).ToNot(BeEmpty())
+			if filepath.Base(path) == "source-controller.crds.yaml" {
+				// this gives us the flux Bucket CRD
+				oneoff = cclist[0]
+			}
+
+			for i, cc := range cclist {
+				// We can't meaningfully check much about converted CRDs in a
+				// generic test like this. But a minimal check is still useful -
+				// props are populated and schemas contain standard k8s metadata
+				// fields at expected paths
+				g.Expect(cc.Props.Spec.Names.Kind).ToNot(BeEmpty(), fmt.Sprintf("elem %d", i))
+				t.Run(cc.Props.Spec.Names.Kind, func(t *testing.T) {
+					NewWithT(t).Expect(cc.Schemas).NotTo(BeEmpty())
+					for _, sch := range cc.Schemas {
+						isch := sch
+						t.Run(sch.Version, func(t *testing.T) {
+							g := NewWithT(t)
+							apiv := isch.Schema.LookupPath(cue.ParsePath("apiVersion?"))
+							g.Expect(apiv.Exists()).To(BeTrue(), fmt.Sprintf("%#v", isch.Schema))
+						})
+					}
+				})
+			}
+		})
+
+		return nil
+	})
+	// TODO does gomega have a better string differ? the output is useless when it fails
+	n, err := format.Node(oneoff.Schemas[0].Schema.Syntax(cue.All(), cue.Docs(true)))
+	g.Expect(err).ToNot(HaveOccurred())
+
+	// hacky one-off test, primarily intended to just let there be _something_
+	// committed to disk that shows what converted output looks like.
+	//
+	// TODO replace with a framework and a test suite that allows testing individual cases of valid constructs within CRD schemas
+	// g.Expect(string(n)).To(Equal(goldenBucketFirstSchema))
+	diff := cmp.Diff(string(n), goldenBucketFirstSchema)
+	if diff != "" {
+		t.Fatal(diff)
+	}
+}
+
+// NOTE - this printed output is created via cue.Value.Syntax() and
+// format.Node() on an original cue.Value that was a definition (#-led label).
+// In such cases, CUE's formatter wraps the output in _#def: {}, because
+// closedness is key to express, but is not an internal property of a struct
+// itself, but the label to which its attached.
+//
+// However, this _#def is ephemerally added only during printing. It does not
+// exist at runtime in the graph; do not look for it with e.g. LookupPath().
+var goldenBucketFirstSchema = `import "strings"
+
+_#def
+_#def: {
+	// APIVersion defines the versioned schema of this representation
+	// of an object. Servers should convert recognized schemas to the
+	// latest internal value, and may reject unrecognized values.
+	// More info:
+	// https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources
+	apiVersion?: string
+
+	// Kind is a string value representing the REST resource this
+	// object represents. Servers may infer this from the endpoint
+	// the client submits requests to. Cannot be updated. In
+	// CamelCase. More info:
+	// https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	kind?: string
+	metadata?: {
+		...
+	}
+
+	// BucketSpec defines the desired state of an S3 compatible bucket
+	spec?: {
```

Comment:
Probably the most important thing to note about this approach in contrast to `cue get go` is that CRDs have all types inlined, and lack the information that would be necessary to recreate that original structure.

So for the `spec` key here, whereas `cue get go` would generate a standalone `#BucketSpec` struct in cue that corresponds to [this struct](https://github.com/fluxcd/source-controller/blob/aa370f284db70e09d8acb49d5bed4e42908b8cb3/api/v1beta1/bucket_types.go#L33), then in the overall resource schema have `spec: #BucketSpec`, here it's all just inlined into one big object.

AFAIK, there's no way around this. My experience over the past couple years tells me that dealing with references like this is generally the most subtle and confounding problem in this general area - one that really benefits from having simple, often restrictive rules.

On the plus side, for this particular case, it's not hard to imagine an approach where the CRD is the baseline that's worked from (since it's always available and most expressive), but if a go package happens to be available, the CLI command could take an additional parameter for the go package, which could then be used to reconstruct the network of types. Handwavy example:

```
timoni mod import crd \
--package flux/source \
-f https://github.com/fluxcd/source-controller/releases/download/v1.0.1/source-controller.crds.yaml \
--go-types github.com/fluxcd/source-controller/api@v1.0.1
```

(the suffix after the `api` package is an interesting question, given that go types are distributed across all those packages)

Reply:
Maybe we could have some conversion function that takes the raw CRD and returns:

```
#Bucket: {
	apiVersion: *"source.toolkit.fluxcd.io/v1beta2" | string
	kind:       *"Bucket" | string

	metadata: {
		name:       string
		namespace?: string
		labels?: {[string]: string}
		annotations?: {[string]: string}
	}

	spec:    #BucketSpec
	status?: #BucketStatus
}

#BucketSpec: {...}

#BucketStatus: {...}
```

CRDs without spec would be consider invalid, only status can be missing. Not sure how to wrap the raw CRD, but I think that the above format should be the `IntermediateCRD` output.

Reply:
Sure, spec and status can be split out generically, and we can set `apiVersion` and `kind`. I'm not sure why you're using a default rather than just a concrete value -  `kind: *"Bucket" | string` rather than just `kind: "Bucket"` - AIUI no other value would be acceptable there.

Reply:
Ah yes, changing the kind doesn’t make sense.